### PR TITLE
CASSANDRA-17466 Shut repair task executor down without interruption to avoid compromising shared channel proxies

### PR DESCRIPTION
**Pull request by maedhroz**

### CODE REVIEW

**Comment by maedhroz** [comment]
https://app.circleci.com/pipelines/github/maedhroz/cassandra?branch=CASSANDRA-17466-trunk&filter=all

File: src/java/org/apache/cassandra/repair/RepairSession.java [hunk 1]
```diff
@@ -372,8 +373,24 @@ public void terminate()
...
+        taskExecutor.shutdown();
         terminate();
+        awaitTaskExecutorTermination();
+    }
+
+    private void awaitTaskExecutorTermination()
+    {
+        try
+        {
+            if (taskExecutor.awaitTermination(30, TimeUnit.SECONDS))
```
Comment (jmckenzie-dev) [inline src/java/org/apache/cassandra/repair/RepairSession.java:385]: What was the reasoning behind the 30 second timeout window here? Might be worth commenting on the selection of this time window.

File: test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java [hunk 2]
```diff
@@ -20,27 +20,37 @@
...
-import org.assertj.core.api.Assertions;
-import org.junit.Test;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
 
 import net.bytebuddy.ByteBuddy;
 import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
 import net.bytebuddy.implementation.MethodDelegation;
+import net.bytebuddy.implementation.bind.annotation.SuperCall;
+import org.apache.cassandra.db.SystemKeyspace;
```
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java]: Looks like import ordering could use some love in this file.

File: test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java [hunk 3]
```diff
@@ -71,6 +81,52 @@ public void testRemoteValidationFailure() throws IOException
...
+            cluster.forEach(i -> i.flush(KEYSPACE));
+            
+            // Flush system.peers_v2, or there won't be any SSTables...
+            cluster.forEach(i -> i.flush("system"));
+            
+            // The remote sync started from node 2 will fail on plan execution and propagate the error...
+            NodeToolResult result = cluster.get(1).nodetoolResult("repair", KEYSPACE);
+            result.asserts().failure().errorContains("Sync failed between /127.0.0.2:7012 and /127.0.0.3:7012");
+
+            assertTrue(cluster.get(1).logs().grep("Stopping transports as disk_failure_policy is stop").getResult().isEmpty());
```
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java:115]: May want to leave a bread crumb comment here (referencing this JIRA + reasoning) as to why we're checking these three specific things to make sure they're not present in the logs.

File: src/java/org/apache/cassandra/streaming/StreamSession.java [hunk 4]
```diff
@@ -183,34 +191,35 @@
...
+ *  |                  |                                        |                 |
+ *  |                  |         if preview                     |                 |
+ *  |                  +----------------------------------------+                 |
+ *  |               nothing to request or to transfer                             |
+ *  +-----------------------------------------------------------------------------+
+ *                  nothing to request or to transfer
+ *
+ *  </pre>
+ */
     public enum State
```
Comment (jmckenzie-dev) [inline src/java/org/apache/cassandra/streaming/StreamSession.java:214]: In general, the fact that we don't actually check our state transitions for validity in StreamSession concerns me a bit. Might want a follow up JIRA to harden state transition logic in this class to where we actually validate it for correctness on state() calls.

Comment (dcapwell) [reply src/java/org/apache/cassandra/streaming/StreamSession.java:214]: +1

File: test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java (hunk 2 above)
Comment (maedhroz) [reply test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java]: Ouch. Yeah, will clean this up.

File: test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java (hunk 3 above)
Comment (maedhroz) [reply test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java:115]: Yeah, I'll comment on how this actually failed before 17466. (i.e. I'll describe the steps to repro as well.)

File: src/java/org/apache/cassandra/repair/RepairSession.java (hunk 1 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/repair/RepairSession.java:385]: The only two things I've actually seen in testing are basically immediate termination and never terminating. To the extent that it identifies both of those things, it's probably fine. The interesting thing would be if abort doesn't clean things up properly, leaving a long-running streaming task open or something. 30 seconds seemed like a long enough time for us to catch that case as well. Are there any other timeouts in the repair/streaming ecosystem that would be good derivatives?

Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/repair/RepairSession.java:385]: I mean, the # seems fine enough to me just wasn't sure where it came from. :)

**Comment by maedhroz** [comment]
Committed as https://github.com/apache/cassandra/commit/7b58b79fe24c612f8bbf7984c325e88496f482d6

**Pull request closed without merging**